ESP32 port, wifi transport, magnetometer, ultrasonic and battery

hippo5329 opened the pull request:
Please pull the following patches,
1. ESP32 port and wdt
2. micro ros wifi transport, OTA, syslog
3. Magnetometer sensors and publisher
4. Ultrasonic sensor and publisher
5. Battery voltage measurement and publisher
6. POSIX clock_gettime

Build and test on ROS2 Humble with esp32dev and Waveshare General Driver for Robots boards.

hippo5329 committed:
port to esp32
  add ServoESP32
  add wrapper for analogWriteFrequency for motor

hippo5329 committed:
use esp32 hardware 64 bits pulse counters for encoders

The esp32 supports hardware 64 bits pulse counters.
No interrupt is required.

hippo5329 committed:
add Syslog support to config

This enables syslog to a remote server with messages like this,

/var/log/linorobot2/hardware.log

2023-05-23T10:25:48.305354-07:00 linorobot2 hardware ﻿setup Ready 4281
2023-05-23T10:25:48.810242-07:00 linorobot2 hardware ﻿createEntities 4788

example server config /etc/rsyslog.conf

# provides UDP syslog reception
module(load="imudp")
input(type="imudp" port="514")

$template Incoming-logs,"/var/log/%HOSTNAME%/%PROGRAMNAME%.log"
*.* ?Incoming-logs

hippo5329 committed:
add wifi transport support to config

hippo5329 committed:
add ArduinoOTA support to config

hippo5329 committed:
update destroy_entities order

Follow the order in upstream,
change the order to destroy executor and node,
https://github.com/micro-ROS/micro_ros_arduino/blob/humble/examples/micro-ros_reconnection_example/micro-ros_reconnection_example.ino

hippo5329 committed:
use upstream I2Cdevlib and remove local copy

Use updated upstream I2Cdevlib and remove local copy.
This fixes some compatibility and stabilility issues.
The MPU6050 and MPU9150 have pgmspace.h issued in upstream.
Only MPU9250 is not in the platformio registry yet. Pulled from
upstream.

hippo5329 committed:
add serial baudrate to config and move serial begin earlier to ease debug

hippo5329 committed:
syncTime current time should be queried after synced with agent

hippo5329 committed:
add AK8963 AK8975 HMC5843 from I2Cdevlib

The AK8963 is not in platformio registry yet.

hippo5329 committed:
add AK09918 code

From
https://github.com/Seeed-Studio/Seeed_ICM20600_AK09918

hippo5329 committed:
add QMI8658 code

https://github.com/hippo5329/QMI8658

hippo5329 committed:
map QMI8658 to IMU interface

hippo5329 committed:
map AK09918 to I2Cdevlib getHeading

hippo5329 committed:
add QMC5883L code

https://github.com/dthain/QMC5883L

hippo5329 committed:
map QMC5883L to I2Cdevlib getHeading

hippo5329 committed:
add mag interface

hippo5329 committed:
add magnetic field publisher

hippo5329 committed:
add hard iron bias

hippo5329 committed:
add orientation

https://github.com/hiwad-aziz/ros2_mpu9250_driver

hippo5329 committed:
add battery

hippo5329 committed:
add battery state publisher

hippo5329 committed:
add i2c pins assignment

hippo5329 committed:
add INA219 current and power sensor

hippo5329 committed:
add namespace prefix to topic

eg, "turtle1/", then the "imu/data" topic will become "turtle1/imu/data".

hippo5329 committed:
change config.h to depend on LINO_BASE so custom config can be added cleanly

hippo5329 committed:
add custom gendrv config for Waveshare General Driver for Robots board

hippo5329 committed:
add ultrasonic sensor HC-SR04. a 5000uS timeout limits range to roughly 0.8m

hippo5329 committed:
add ultrasonic range sensor publisher

hippo5329 committed:
add esp32 to calibration

hippo5329 committed:
add serial baudrate config to calibration

hippo5329 committed:
add BOARD_INIT to calibration for board specific setup

hippo5329 committed:
add BOARD_INIT for board specific setup

hippo5329 committed:
add short brake for shorter stopping distance

hippo5329 committed:
use POSIX clock_gettime() when available
to mitigate the overflow issue of millis() in 50 days.
change syncTime timeout to 1000mS as it was in the micro-ROS tutorial

hippo5329 committed:
add esp32 task watchdog

hippo5329 committed:
fullStop on agent dissconnected

hippo5329 committed:
allow customized RCCHECK and use syslog to report RCCHECK failure

hippo5329 committed:
Update README.md for ESP32 port and some devices

hippo5329 committed:
Update README.md battery and echo input voltage divider limit to 3.3V

hippo5329 committed:
Update README.md preferred custom configuration

hippo5329 committed:
motor skip unused pins

hippo5329 committed:
add diagnosis utility test_motors

Spin the motors forward and backward alternately.
And display the motors speed.

MOTOR0 FWD RPM    156.3     -4.6      0.0      0.0
...
MOTOR0 FWD RPM    161.2      0.0      0.0      0.0
MOTOR1 FWD RPM      3.7    165.5      0.0      0.0
...
MOTOR1 FWD RPM      0.0    170.6      0.0      0.0
MOTOR0 REV RPM   -160.1      4.8      0.0      0.0
...
MOTOR0 REV RPM   -165.1      0.0      0.0      0.0
MOTOR1 REV RPM     -3.9   -161.1      0.0      0.0
...
MOTOR1 REV RPM      0.0   -167.2      0.0      0.0

hippo5329 committed:
add diagnosis utility test_sensors

Display IMU, MAG, battery and range sensors data.

ACC -0.77  0.69  9.47 GYR  0.00 -0.01  0.00 MAG -239.00 -326.00 -3.00
 BAT 12.59V RANGE  0.22m

hippo5329 committed:
Update README.md diagnosis utilities for motors and sensors

hippo5329 committed:
use short brake aka fast decay in BTS7960 motor drive for better control

hippo5329 committed:
test_motors fix motor names

hippo5329 committed:
disable orientation with MAG because it interferes with EKF

hippo5329 committed:
use FAKE_IMU when no real IMU defined

hippo5329 committed:
define NODE_NAME

hippo5329 committed:
push lidar data to UDP server tested with LD19

hippo5329 committed:
update base config and gendrv for lidar etc

hippo5329 committed:
update README on customization and remove the Waveshare gendrv

hippo5329 committed:
Revert "disable orientation with MAG because it interferes with EKF"

This reverts commit 6289547f95fdd0e9868eb80a7020f31bb0f12a2e.
The problem was caused by IMU calibration during setup().
It needs to be still.

hippo5329 committed:
base config suggested BOARD_INIT to wait before IMU calibration

The robot needs to be still to do IMU calibration.

hippo5329 committed:
lidar use serial callback to reduce lost data

hippo5329 committed:
remove wifi initialization code from set_microros_wifi_transports

wifi will be indepent of micro ros transport
wifi is enabled by defining WIFI_AP_LIST
support multiple wifi ap
when wifi signal is too low, disconnect current ap and
scan for the strongest signal

hippo5329 committed:
lidar serial begin after parameters set

hippo5329 committed:
Update README.md  WIFI and LIDAR config

hippo5329 committed:
calibration remove esp32 from project

The calibration utility does not work on esp32.
Please use test_motors and test_sensors instead.

hippo5329 committed:
Update README.md esp32 wifi transport and lidar udp transport

hippo5329 committed:
move wifi code to lib

hippo5329 committed:
move ArduinoOTA to lib/wifi

hippo5329 committed:
test_motors add wifi ota syslog

hippo5329 committed:
test_sensors add wifi ota syslog

hippo5329 committed:
wifi add syslog show ip when wifi connected

hippo5329 committed:
README update esp32 wifi usage

hippo5329 committed:
config make wifi setting consistent with RAEDME

hippo5329 committed:
config change esp32 flash partitions for larger app space

hippo5329 committed:
test_motors display speed and stopping distance

hippo5329 committed:
README move calibration to calibration project directory
and update test_motors speed and stopping distance

hippo5329 committed:
add pico
initial commit, compiled but not tested on hardware yet.

hippo5329 committed:
add pico-w

hippo5329 committed:
ota change for picow

hippo5329 committed:
encoder pico update
The pull request is merged upstream.

grassjelly committed:
removed unused robots

atticusrussell commented:
@hippo5329 discussing how to go about merging this PR in #54 and would love your input.

hippo5329 committed:
esp32: change Servo library

The previous Servo library from RoboticsBrno/ServoESP32 failed to compile,
so we switched to madhephaestus/ESP32Servo. As it was sugggested by
robofounddry in
https://github.com/robofoundry/linorobot2_hardware_hippo_esp32_fix

hippo5329 committed:
wifi: change include WiFi.h position
to avoid errors on platforms without Wifi support

hippo5329 committed:
board: add waveshare general driver for robots

hippo5329 committed:
board: add esp32 diff drive tank

hippo5329 committed:
README: markdown lint and fix style

hippo5329 committed:
.gitignore: ignore .vscode

hippo5329 committed:
README: add vscode/platformio ide for esp32

hippo5329 committed:
README: move vscode to the beginning of software installation

hippo5329 committed:
RAEDME: add robot hardware

hippo5329 committed:
README: move documentation to Wiki

hippo5329 committed:
board: revert to usb for initial download

hippo5329 committed:
RAEDME: add vscode installation command lines

hippo5329 committed:
config: remove monitor_speed and comment out BAUDRATE

hippo5329 committed:
rename firmware.ino to firmware.cpp and reorder func call sequences

otherwise the build will generate firmware.ino.cpp
the compilation errors cannot be tracked.

after renamed, the func must be declared before uses. so the func codes are reordered.

hippo5329 committed:
lib/mag: remove unused variable

hippo5329 committed:
RAEDME: remove brltty deb pkg for CH340 uart chip

hippo5329 committed:
firmware: debug msg on IMU init failed

hippo5329 committed:
config: esp32tank PWM_BITS changed to 10

hippo5329 committed:
config: use micro_ros_platformio default serial iron unless specified

hippo5329 committed:
README: remove pico-w and fix typo

hippo5329 committed:
README: why esp32

hippo5329 committed:
README: revert to platformio cli

hippo5329 committed:
README: connect the micro ros agent wifi transport

hippo5329 committed:
board: add esp32meca  Mecanum drive robot

hippo5329 committed:
config: revert minitor_speed 0d15f6d and define BAUDRATE 115200

hippo5329 committed:
RAEDME: add supported details

hippo5329 committed:
RAEDME: 4wd and wifi ip

hippo5329 committed:
README: typo micro-ROS HC-SR04

hippo5329 committed:
battery: add adc_calibrate utility

hippo5329 committed:
battery: default no USE_ADC_LUT

hippo5329 committed:
adc_calibrate: use firmware platformio.ini

But use of wifi failed dram usage.

hippo5329 committed:
imu: use I2Cdevlib-MPU6050

hippo5329 committed:
Revert "config: use micro_ros_platformio default serial iron unless specified"

This reverts commit 454a0b8d09f221782fca18b3cd40b3043ca4ed5c.

It is better to specify ROS2 distribution in use.

hippo5329 committed:
config: remove esp32tank and esp32meca use esp32_config.h instead

hippo5329 commented:
prepare new pr
